test/lib/pwa.js

```diff
@@ -32,39 +36,114 @@ var assert = require('chai').assert;
+      simpleMock.mock(libPwa, 'db').returnWith(db);
+      return libPwa.updateLighthouseInfo(pwa).should.be.fulfilled.then(updatedPwa => {
+        assert.equal(libLighthouse.fetchAndSave.callCount, 1);
+        assert.equal(libLighthouse.fetchAndSave.lastCall.args[0], '123456789');
+        assert.equal(db.update.callCount, 1);
+        assert.equal(updatedPwa.lighthouseScore, 82);
+      });
+    });
+  });
 
-          // Mock libImages and bd to avoid making real calls
-          simpleMock.mock(libImages, 'fetchAndSave').resolveWith(manifest.getBestIconUrl());
-          pwaLib.libImages = libImages;
-          simpleMock.mock(db, 'update').returnWith(pwa);
-          pwaLib.db = db;
+  describe('#save (core logic)', () => {
+    afterEach(() => {
+      simpleMock.restore();
+    });
+    it('performs all the save steps', () => {
+      // Mock findByManifestUrl and bd to avoid making real calls
+      simpleMock.mock(libPwa, 'findByManifestUrl').resolveWith(pwa);
+      simpleMock.mock(libManifest, 'fetchManifest').resolveWith(manifest);
+      simpleMock.mock(libPwa, 'fetchMetadataDescription').resolveWith('test-description');
+      simpleMock.mock(db, 'update').returnWith(pwa);
+      simpleMock.mock(libPwa, 'updateIcon').resolveWith(pwa);
+      simpleMock.mock(libPwa, 'updateLighthouseInfo').resolveWith(pwa);
+      simpleMock.mock(libPwa, 'libManifest').returnWith(libManifest);
+      simpleMock.mock(libPwa, 'db').returnWith(db);
 
-          var promiseUpdateIcon = pwaLib.updateIcon(pwa, manifest);
-          assert.equal(libImages.fetchAndSave.callCount, 1);
-          assert.equal(libImages.fetchAndSave.lastCall.args[0], 'https://s1.trrsf.com/fe/zaz-morph/_img/launcher-icon.png?v2');
-          assert.equal(libImages.fetchAndSave.lastCall.args[1], '123456789.png');
-          return promiseUpdateIcon.should.eventually.have.property('iconUrl');
-        });
+      return libPwa._save(pwa).should.be.fulfilled.then(updatedPwa => {
+        assert.equal(libPwa.findByManifestUrl.callCount, 1);
+        assert.equal(libPwa.findByManifestUrl.lastCall.args[0], MANIFEST_URL);
+        assert.equal(libManifest.fetchManifest.callCount, 1);
+        assert.equal(libManifest.fetchManifest.lastCall.args[0], MANIFEST_URL);
+        assert.equal(updatedPwa.manifest, manifest);
+        assert.equal(libPwa.fetchMetadataDescription.callCount, 1);
+        assert.equal(libPwa.fetchMetadataDescription.lastCall.args[0], pwa);
+        assert.equal(updatedPwa.metaDescription, 'test-description');
+        assert.equal(db.update.callCount, 1);
+        assert.equal(libPwa.updateIcon.callCount, 1);
+        assert.equal(libPwa.updateLighthouseInfo.callCount, 1);
+      });
+    });
+    it('handles E_MANIFEST_ERROR error', () => {
+      // Mock findByManifestUrl to avoid making real calls
+      simpleMock.mock(libPwa, 'findByManifestUrl').rejectWith(new Error('Testing error'));
```

Comment:
@ithinkihaveacat let me move it to another place. I can also add the error codes here.